projects/RabbitMQ.Client.OAuth2/OAuth2Client.cs

```diff
@@ -169,127 +114,162 @@ internal class OAuth2Client : IOAuth2Client, IDisposable
-                IDictionary<string, string> additionalRequestParameters,
-                HttpClientHandler httpClientHandler)
+        public OAuth2Client(string clientId, string clientSecret, Uri tokenEndpoint,
+            string? scope,
+            IDictionary<string, string>? additionalRequestParameters,
+            HttpClientHandler? httpClientHandler)
         {
-            this._clientId = clientId;
-            this._clientSecret = clientSecret;
-            this._scope = scope;
-            this._additionalRequestParameters = additionalRequestParameters == null ? EMPTY : additionalRequestParameters;
-            this._tokenEndpoint = tokenEndpoint;
+            _clientId = clientId;
+            _clientSecret = clientSecret;
+            _scope = scope;
+            _additionalRequestParameters = additionalRequestParameters == null ? EMPTY : additionalRequestParameters;
+            _tokenEndpoint = tokenEndpoint;
 
             _httpClient = httpClientHandler == null ? new HttpClient() :
                 new HttpClient(httpClientHandler);
             _httpClient.DefaultRequestHeaders.Accept.Clear();
             _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
         }
 
-        public IToken RequestToken()
+        public async Task<IToken> RequestTokenAsync(CancellationToken cancellationToken = default)
         {
-            var req = new HttpRequestMessage(HttpMethod.Post, _tokenEndpoint);
-            req.Content = new FormUrlEncodedContent(buildRequestParameters());
-
-            Task<HttpResponseMessage> response = _httpClient.SendAsync(req);
-            response.Wait();
-            response.Result.EnsureSuccessStatusCode();
-            Task<JsonToken> token = response.Result.Content.ReadFromJsonAsync<JsonToken>();
-            token.Wait();
-            return new Token(token.Result);
+            using HttpRequestMessage req = new HttpRequestMessage(HttpMethod.Post, _tokenEndpoint);
+            req.Content = new FormUrlEncodedContent(BuildRequestParameters());
+
+            using HttpResponseMessage response = await _httpClient.SendAsync(req)
+                .ConfigureAwait(false);
+
+            response.EnsureSuccessStatusCode();
+
+            JsonToken? token = await response.Content.ReadFromJsonAsync<JsonToken>()
+                .ConfigureAwait(false);
+
+            if (token is null)
+            {
+                // TODO specific exception?
+                throw new InvalidOperationException("token is null");
+            }
+            else
+            {
+                return new Token(token);
+            }
         }
 
-        public IToken RefreshToken(IToken token)
+        public async Task<IToken> RefreshTokenAsync(IToken token,
+            CancellationToken cancellationToken = default)
         {
             if (token.RefreshToken == null)
```

Comment:
There is a nullable mismatch here, you check for ``RefreshToken == null``, but in the interface ``IToken`` the property is not nullable, as having a token without a refresh_token is perfectly fine, I'd recommend making the property in ``IToken`` nullable.